### PR TITLE
[MXNET-857] Add initial NVTX profiler implementation

### PR DESCRIPTION
**Pull request by KellenSunderland**
## Description ##
This PR builds on-top of current profiler support to allow profiling via NVIDIA's NVTX APIs.  These extensions mark readable ranges in the NVIDIA Visual Profiler which helps show correlations between kernel launches and graph node executions.

Example shown here: https://user-images.githubusercontent.com/7443219/33946110-34296d18-e021-11e7-8d18-6d40b797405c.png  (The additional information enabled is in the 'Markers and Ranges' row.)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this changs

Notes:
* TODO: Integration test

### CODE REVIEW

File: tests/python/profiling/test_nvtx.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+def test_nvtx_ranges_present_in_profile():
+
+    if not mx.test_utils.list_gpus():
+        unittest.skip('Test only applicable to machines with GPUs')
+
+    print(sys.executable)
+    
+    print(os.path.realpath(__file__))
+
```
Comment (KellenSunderland) [inline tests/python/profiling/test_nvtx.py]: TODO: launch subprocess that profiles the simple forward pass of network py.  Verify that ranges are collected properly.

**Comment by ptrendx** [comment]
One thing that would make the profile more readable is different colors for different ranges. Here is an example how to do this: https://github.com/NVIDIA/DALI/blob/master/dali/common.h#L149

You could for example use the first letter of the range as a very simple hash to the list of predetermined colors (don't do anything more fancy like actual hashing of the name since that would only introduce overhead which you don't want during profiling).

**Comment by KellenSunderland** [comment]
Great idea.  Totally agree.

**Comment by stu1130** [comment]
@mxnet-label-bot [pr-work-in-progress]

**Comment by vandanavk** [comment]
@KellenSunderland Is the PR still WIP?

**Comment by KellenSunderland** [comment]
@vandanavk Yes, I'll loop around to this once I finish some other tasks.  Does it cause problems for your team to leave it in a WIP state?  For me it's quite handy to be able to have WIP PRs that I can put aside and resume later.

**Comment by Roshrini** [comment]
@KellenSunderland It doesn't cause problems for our team. But we do try to go through all the stale PRs in the repo to make sure someone is working on it and PRs are getting timely reviewed.

**Comment by lupesko** [comment]
@KellenSunderland thanks for the contribution! Any updates on progress/ETA?

**Comment by KellenSunderland** [comment]
Hey @lupesko.  Sorry still in my backlog.  I've reviewed a few suggestions so no longer blocked on anyone.  I just need to find the time to finish this off.

**Comment by anirudhacharya** [comment]
@KellenSunderland pinging for update.

**Comment by stu1130** [comment]
@KellenSunderland ping
feel free to close the PR and reopen it if you have time :)

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-awaiting-testing]

**Comment by roywei** [comment]
@KellenSunderland ping for update and trigger ci. thanks!

**Comment by sandeep-krishnamurthy** [comment]
This will be very helpful addition to profiler utility. @KellenSunderland - looking forward for the update.

**Comment by pinaraws** [comment]
@mxnet-label-bot add[pr-awaiting-review]

**Comment by pinaraws** [comment]
@mxnet-label-bot remove[pr-work-in-progress]

**Comment by pinaraws** [comment]
@rahulhuilgol Please review

**Comment by abhinavs95** [comment]
@anirudh2290 @szha Could you please help review?

**Comment by piyushghai** [comment]
@anirudh2290 Ping again.

**Comment by lupesko** [comment]
@KellenSunderland is this still WIP? name says it is, label says otherwise

File: tests/python/profiling/test_nvtx.py (hunk 1 above)
Comment (KellenSunderland) [reply tests/python/profiling/test_nvtx.py]: Todo done.  Testing in CI to make sure it correctly detects ranges.

**Comment by KellenSunderland** [comment]
Still making progress on this.  I believe the last TODO is either use static linking or copy make sure the nvExt library is copied to the lib path for testing on windows.  It's probably best for the customer to use static linking, so I'll try to implement that.

**Comment by KellenSunderland** [comment]
@sandeep-krishnamurthy ready for review if you've got a second.

File: tests/python/profiling/simple_forward.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import mxnet as mx
+from mxnet.gluon import nn
+
+
+def simple_forward():
```
Comment (eric-haibin-lin) [inline tests/python/profiling/simple_forward.py]: Is there any plan to add some documentation on how to use this feature as an user?

Comment (KellenSunderland) [reply tests/python/profiling/simple_forward.py]: So there's no change required from using the profiler as it exists now, you'll just get more data when using nvprof.

That being said I do want to create documentation specifically describing how to profile with nvprof and the profiler, and describing what kind of output to expect.  I'm ok with deferring this PR until that one is available if you'd prefer.

Comment (KellenSunderland) [reply tests/python/profiling/simple_forward.py]: @eric-haibin-lin Added some more information into the existing profiler tutorials.

Comment (eric-haibin-lin) [reply tests/python/profiling/simple_forward.py]: Looks good!

**Comment by eric-haibin-lin** [comment]
LGTM. Will this be enabled by default (e.g. via pip)?

**Comment by KellenSunderland** [comment]
Yes, it'll be on by default for CUDA builds.

**Comment by KellenSunderland** [comment]
@eric-haibin-lin Any other questions about this PR?  I've done a bunch of perf tests offline and don't see any regressions, but don't mind doing some additional investigations if you'd like.  If you're ok with the change would you be able to set this PR to approved, I'm hesitant to merge without green checkboxes :-).

File: CMakeLists.txt [hunk 3]
```diff
@@ -285,6 +285,16 @@ if(USE_CUDA)
       message(WARNING "Could not find NCCL libraries")
     endif()
   endif()
+  if(UNIX)
+    find_package(NVTX)
+    if(NVTX_FOUND)
+      include_directories(${NVTX_INCLUDE_DIRS})
+      list(APPEND mxnet_LINKER_LIBS ${NVTX_LIBRARIES})
+      add_definitions(-DMXNET_USE_NVTX=1)
```
Comment (szha) [inline CMakeLists.txt]: can you add support in makefile too?

Comment (KellenSunderland) [reply CMakeLists.txt]: Can do.

**Comment by eric-haibin-lin** [comment]
@KellenSunderland any update? look forward to this feature

**Comment by KellenSunderland** [comment]
Thanks @eric-haibin-lin.  Updated with a Makefile build.  Sorry it took a while, just wanted to make sure I tested it well with that build method.

File: make/config.mk [hunk 4]
```diff
@@ -80,6 +80,9 @@ ENABLE_CUDA_RTC = 1
 # whether use CuDNN R3 library
 USE_CUDNN = 0
 
+# whether to use NVTX when profiling
+USE_NVTX = 0
```
Comment (szha) [inline make/config.mk]: @KellenSunderland do you recommend enabling this in pip?

Comment (KellenSunderland) [reply make/config.mk]: I did measurements with a few models and didn't see any performance deltas.  I think it would be safe to enable in pip.

Comment (ptrendx) [reply make/config.mk]: Yes, as long as you are not running under nvprof nvtx calls are basically noops.

**Pull request merged**